src/torchmetrics/functional/regression/kendall.py

```diff
@@ -0,0 +1,396 @@
+        min_classes = torch.minimum(preds_unique, target_unique)
+        tau = 2 * con_min_dis_pairs / ((min_classes - 1) / min_classes * n_total**2)
+
+    return tau
+
+
+def _calculate_p_value(
+    con_min_dis_pairs: Tensor,
+    n_total: Tensor,
+    preds_ties: Optional[Tensor],
+    preds_ties_p1: Optional[Tensor],
+    preds_ties_p2: Optional[Tensor],
+    target_ties: Optional[Tensor],
+    target_ties_p1: Optional[Tensor],
+    target_ties_p2: Optional[Tensor],
+    variant: _MetricVariant,
+    alternative: Optional[_TestAlternative],
+) -> Tensor:
+    """Calculate p-value for Kendall's tau from metric metadata."""
+    normal_dist = torch.distributions.normal.Normal(torch.tensor([0.0]), torch.tensor([1.0]))
+
+    t_value_denominator_base = n_total * (n_total - 1) * (2 * n_total + 5)
+    if variant == _MetricVariant.A:
+        t_value = 3 * con_min_dis_pairs / torch.sqrt(t_value_denominator_base / 2)
+    else:
+        m = n_total * (n_total - 1)
+        t_value_denominator: Tensor = (t_value_denominator_base - preds_ties_p2 - target_ties_p2) / 18
+        t_value_denominator += (2 * preds_ties * target_ties) / m  # type: ignore (is Tensor)
+        t_value_denominator += preds_ties_p1 * target_ties_p1 / (9 * m * (n_total - 2))  # type: ignore (is Tensor)
+        t_value = con_min_dis_pairs / torch.sqrt(t_value_denominator)
+
+    if alternative == _TestAlternative.TWO_SIDED:
+        t_value = torch.abs(t_value)
+    if alternative in [_TestAlternative.TWO_SIDED, _TestAlternative.GREATER]:
+        t_value *= -1
+    p_value = normal_dist.cdf(t_value)
+    if alternative == _TestAlternative.TWO_SIDED:
+        p_value *= 2
+    return p_value
+
+
+def _kendall_corrcoef_update(
+    preds: Tensor,
+    target: Tensor,
+    concat_preds: List[Tensor] = [],
+    concat_target: List[Tensor] = [],
+    num_outputs: int = 1,
+) -> Tuple[List[Tensor], List[Tensor]]:
+    """Update variables required to compute Kendall rank correlation coefficient.
+
+    Args:
+        preds: Ordered sequence of data
+        target: Ordered sequence of data
+        concat_preds: List of batches of preds sequence to be concatenated
+        concat_target: List of batches of target sequence to be concatenated
+        num_outputs: Number of outputs in multioutput setting
+
+    Raises:
+        RuntimeError: If ``preds`` and ``target`` do not have the same shape
+    """
+    # Data checking
+    _check_same_shape(preds, target)
+    if preds.ndim > 2 or target.ndim > 2:
+        raise ValueError(
+            f"Expected both predictions and target to be either 1- or 2-dimensional tensors,"
+            f" but got {target.ndim} and {preds.ndim}."
+        )
+    if (num_outputs == 1 and preds.ndim != 1) or (num_outputs > 1 and num_outputs != preds.shape[1]):
+        raise ValueError(
+            f"Expected argument `num_outputs` to match the second dimension of input, but got {num_outputs}"
+            f" and {preds.shape[1]}."
+        )
```

Comment:
maybe at this point it is worth refactoring this into an utility function when, since it is repeated between all three correlation metrics